artiq/device_db_testing/device_db_generation.py

```diff
@@ -0,0 +1,139 @@
+                    }
+            }
+        })
+
+    return dio
+
+#Return dictionary with information about urukul device
+#in format required for device_db file.
+def add_urukul(devices_config, device_id, custom_info, generate_all):
+    device_name = custom_info.pop("name", None)
+    if device_name == None:
+        device_name = device_id
+
+    urukul = {
+            device_name+"_spi": {
+                            "type": "local",
+                            "module": "artiq.coredevice.spi2",
+                            "class": devices_config[device_id]["rtio_channels"]["spi"]["class"],
+                            "arguments": {
+                                    "channel": devices_config[device_id]["rtio_channels"]["spi"]["channel_num"]
+                            }
+            },
+
+            device_name+"_ttl_io_update": {
+                            "type": "local",
+                            "module": "artiq.coredevice.ttl",
+                            "class": devices_config[device_id]["rtio_channels"]["io_update"]["class"],
+                            "arguments": {
+                                    "channel": devices_config[device_id]["rtio_channels"]["io_update"]["channel_num"]
+                            }
+            },
+
+            device_name+"_cpld": {
+                            "type": "local",
+                            "module": "artiq.coredevice.urukul",
+                            "class": "CPLD",
+                            "arguments": {
+                                "spi_device": device_name+"_spi",
+                                "io_update_device": device_name+"_ttl_io_update",
+                                "refclk": 100e6,
+                                "clk_sel": 1
+                            }
+            }
+    }
+
+    for channel_id in devices_config[device_id]["rtio_channels"]:
+        if channel_id[-1].isdigit():
+            for (channel, custom_name) in custom_info.items():
+                if channel_id[-1] in str(channel):
+                    name = custom_name
+                    break
+            else:
+                if generate_all == True:
+                    name = device_name+"_ch"+channel_id[-1]
+                else:
+                    continue
+
+            urukul.update({
+                device_name+"_ttl_"+channel_id: {
+                                "type": "local",
+                                "module": "artiq.coredevice.ttl",
+                                "class": devices_config[device_id]["rtio_channels"][channel_id]["class"],
+                                "arguments": {
+                                        "channel": devices_config[device_id]["rtio_channels"][channel_id]["channel_num"]
+                        }
+                },
+
+                name: {
+                        "type": "local",
+                        "module": "artiq.coredevice.ad9912",
+                        "class": "AD9912",

```

Comment:
We need a parameter to allow us to choose either the AD9910 or AD9912 versions of Urukul (there are two hw revisions with different DDS chips on).